aten/src/ATen/native/cuda/BatchLinearAlgebraLib.h

```diff
@@ -12,13 +12,54 @@
 #define USE_CUSOLVER
 #endif
 
+// comment out this line as the emergency brake if there is something wrong with parallel launch
+// e.g. if you have a script failure with linear algebra and set environment variable CUDA_LAUNCH_BLOCKING=1
+// solves the problem, then you can try removing the cuda parallel stream launch
+#define _USE_PARALLEL_LAUNCH
+
+
+#ifdef _USE_PARALLEL_LAUNCH
+
+// Launch multiple operations in different CUDA stream in parallel, using at::parallel_for to speed up.
+#define CUDA_PARALLEL_STREAM_LAUNCH(_i, _batch_size, _stmt)                         \
```

Comment:
This whole macro might be a very controversial change, so I'm pointing it out in advance. It basically does two things:

1. The automatic recording and blocking of cuda stream and event.
2.  `at::parallel_for` to launch kernels in parallel much faster than serial for-loop launch.

Point 1 is mainly for the convenience of code writing. 

Point 2 improves the performance on A100 and V100 by a lot. See benchmark links here https://github.com/pytorch/pytorch/pull/48436#issuecomment-733430414. On my 2070 super, the performance improves a bit for some shapes.

Point 2, the division by 4 is a heuristic. I tested that with different divisor here https://github.com/xwang233/code-snippet/tree/master/svd/at-parallel-for

Point 2, this change would also affect `torch.inverse` with batch size 2. Fortunately, the performance of that basically remains the same. On my 2070 super, 2x512x512 gets improved from 3.53 ms to 2.95 ms. 2x1024x1024 gets regressed from 12.2 ms to 14.4 ms.

Reply:
Why does at::parallel_for help compared to for-loop launch? Is it because of the threading? If so, that's unsafe - device is correctly set with deviceGuard only on the main thread, all extra threads will have the wrong device.

Reply:
Can you modify your benchmarks to limit the number of cpu cores you are using with `taskset`? If you are (potentially) using all the CPU cores on the powerful CPUs on DGX1Vs and Ampere machines to drive a single GPU, that's not realistic.

Reply:
I'll add a cuda guard to `at::parallel_for`, and seems like it fixes the problem for wrong device on my machine so far.

I think this setup for grain size can limit the number of parallelism to a maximum of 4?
```cpp
 at::parallel_for(0, _batch_size, _batch_size / 4 + 1, [&](int _start, int _end){
```

Reply:
There aren't many performance gain after cuda guard is added, and there are some weird behavior as well. So I reverted the `at::parallel_for` changes.